### ISSUE TITLE
Epic: Evacuator evolution

### ISSUE THREAD

**Issue by ktbriggs** [eclipse/omr#2664]
This is a container for reporting investigation of prospective avenues of further development of the stack-based traversal algorithm developed for evacuator (#2361). Stable source code the evacuator algorithm can be found in the `evacuator` branch of my public orm and openj9 repos:

https://github.com/ktbriggs/omr/tree/evacuator (see `gc/base/Evacuator*.*pp`)
https://github.com/ktbriggs/openj9/tree/evacuator (see `runtime/gc_glue_java/Evacuator*.*pp`)

The items below will be developed and benchmarked and reviewed here as evacuator development continues. 

**Notice:** (Jan 29 2019) I am wrapping up work on evacuator for now and have added a `Status` comment for each of the items below. I will soon squash all commits subsequent to the base `OMR Integration` commit into one `OpenJ9 Integration` commit. The former commit has been through the Eclipse CQ process and previously reviewed.

-- _interleave survivor and tenure scanspaces on the scan stack_

Evacuators currently maintain homogeneous scanspaces on the stack -- once work packet with survivor or tenure scan work is loaded into the bottom frame, subsequent pushes involve only objects copied to the same region and all objects evacuated to the other region are copied to outside copyspaces. This change will allow any small object to be pushed up the stack. This is especially important for remembered tenured arrays of small objects (eg java String). With this change all array elements remaining in survivor space will be copied inside the stack and collocated with their surviving or tenured referents. 

This change is necessary and moderately difficult (2-3 weeks work).

Status: Completed: see benchmarking results below

-- _push root, remembered and clearable objects onto the scan stack_

Evacuated roots, referents of remembered objects and clearable objects are currently copied breadth first to outside copyspaces. This change will push each such object onto the (empty) scan stack and spin their referents out on the stack until to stack pops to empty. This change leverages interleaved scanspaces and is expected to further improve object collocation in survivor and tenure regions.

This change is necessary and not difficult (3-5 days).

Status: Completed: see benchmarking results below

-- _prioritize scanning of work in tenure region_

Likelihood of referent object collocation when scanning inside a stack frame decreases over time because frequency of encountering already forwarded referents increases as nursery collection proceeds. This will pull tenure work from evacuator work lists before pulling any survivor work which should improve referent object collocation in the tenure region.

This change is to be tested and discarded if not found to improve performance in tenure-bound workloads like SPECjvm2008/derby. It is not difficult to implement and test (2-3 days).

Status: Abandoned: no advantage found in any salient metric

-- _conditionally include evacuator source code_

As required to cleanly separate evacuator development from release stream while producing/enabling builds of evacuator+scavenger-enabled SDKs. 

This change is necessary and somewhat difficult (1-2 weeks).

Status: Partially completed: Scavenger.cpp has been lightly modified to dispatch gc threads to run `MM_Evacuator::workThreadGarbageCollect()` instead of `MM_Scavenger::workThreadGarbageCollect` in evacuator builds only when `-Xgc:recursiveScanOrdering` is explicitly selected as a command-line option. Scavenger operation and performance are in evacuator builds essentially identical to production builds when this option is excluded from the command line.

-- _incorporate stack-based scanning and copying into the balanced collector_

The balanced collector model generalizes the standard nursery collector model by extending the number of evacuation spaces from 3 (evacuate/survivor/tenure) to 25 (heap region age 0-24). The gencon evacuator model can be similarly generalized to N evacuation spaces that can be interleaved on the scan stack. and could employ the evacuator scanning model. This change will involve generalizing the evacuator model as required to enable balanced integration while maintaining gencon integration.

Goal is to produce one evacuation engine that can be used with gencon or balanced policies. This is a big deal and difficult (3-4 months). 

Status: Not attempted.

-- _recognize and chain recursive structures_

The evacuator stack seldom overflows expect when a recursive structure (eg, linked list) is being scanned. For applications with deep recursing chains where elements have substantial payloads of mostly small objects this can become problematic for the processing evacuator as it may spend an inordinate amount of time processing chains in the structure without generating much outside copy. Evacuator can recognize this condition (stack overflow) and flush the stack, examining every scanned reference for a recursive relationship with parent (class equivalence). If a recursive reference is found, it can then be scanned and copied in a manner that copies recursing referents inside the stack and copies other referents to outside copy spaces, thereby distributing the work of processing these chains.

This is an important change for applications that use recursive structures involving long reference chains (eg, SPECjvm2008/xml.validation, SPECjbb2015). Plots of nursery collection times for these applications show large spikes and these have been identified as long linked lists with substantial referent material in each linked node. This change should eliminate these spikes. Moderately difficult (2-3 weeks), as time and priorities permit.

Status: Partially completed: The original idea behind this was to use stack overflow as signal that a recursive structure is being scanned. In that case, it was believed that a recurring link could be detected by 1) setting inside copy distance = 0 when the stack pointer returned to bottom frame (to force depth-first scanning) and 2) check the linear chain of objects being scanned for sub/superclass relationships if the stack overflowed again. This would be viable except for the common case (eg, XML DOM) when the recurring links are embedded in classes that have a common superclass but no direct sub/superclass relationship. 

Instead, this is partially addressed by 1) setting inside copy distance = 0 when the stack pointer returns to bottom frame after stack overflow and then flushing all inferior frames to outside copy spaces as the stack unwinds. For branching recursive structures like XML DOM, this will push all but the first element scanned in each frame outside, and these elements may include other recursing chains that would otherwise be copied inside that stack. Some of this outside work may be distributable to other evacuator threads.

The impact of this change is reflected in improved cpu utilization stats for evacuator in "final" benchmarking results below. 

-- _copy objects depth-first on the scan stack_

The current evacuator algorithm uses the same aliasing technique as scavenger, but more successfully as it constrains copy-scan distance to the width of a stack frame (~4kb). As a result only about ~20% of all copy operations result in 64 byte cache-line containment of referring pointer and head of copied referent, because the copy-scan distance tends to increase as the scan head moves away from the base of the frame. If referent objects were always copied up to the next stack frame, then the first pushed referent from any object would be located contiguously with its parent object. It is expected that this change would improve cache-line containment from ~20% to 40-50% and, when combined with hot field selection for first scanned referent (see #2663), this could result in a measurable boost to application throughput.

This experimental, not planned, and moderately difficult (2-3 weeks).

Status: Completed: Depth-first scanning is obtained when `-XXgc:recursiveMaximumInsideCopyDistance` is set to 0 on the command line. The results are ambiguous -- this mode of operation typically increases cache-line containment into the 30-50% range (2-3X greater than with default operational parameters), but the impact on application throughput is inconsistent and difficult to separate from reduced GC throughput due to additional stack overhead. This feature could be greatly improved if JIT could provide information regarding "hotness" of fields per class, in which case object scanners could be tuned to return the hottest reference slot first and the referenced "hot" field would then be copied adjacent to the referring object. See benchmarking results below.

-- _apply stack-based depth-first scanning algorithm to standard marking_

The standard marking scheme uses a breadth-first algorithm to mark and enqueue marked objects for scanning. If the nursery collector uses a stack-based depth-first algorithm for evacuating objects to survivor and tenure space, the frequent close proximity of parent and referent objects can be leveraged to greater effect if the marking algorithm also uses a stack-based depth-first algorithm. The performance boost for marking would them accrue from selecting a proximate referent object for scanning and from reduced trafficking in work packets (since most marking and scanning would occur on the stack). 

This is just a thought, unplanned.

Status: Not attempted: I do believe that marking performance would benefit from stack-based scanning. In fact, I am convinced that improved Evacuator and Scavenger performance vs breadth-first copying is mainly due to reduced trafficking of work units between gc threads, and this advantage would accrue to marking as well if a similar stack-based traversal were to be employed. 

-- _replace single-threaded heap scan in MM_PhysicalSubArenaVirtualMemorySemiSpace::contract()_

This method walks the nursery to locate and map addresses affected by nursery contraction and is single-threaded because the mark map may not be up to date at time of contraction. This can be parallelized as for evacuator by recursively walking objects starting from root and remembered set.  

This is just a thought, unplanned.

Status: Not attempted.

**Comment by ktbriggs** [comment]
Resurrected and continuing evacuator development here: https://github.com/ktbriggs/omr/tree/evacuator. This branch includes previously reviewed commit `OMR Integration` unchanged modulo rebasing and antecedent commits relating to development of this epic. It is continuously rebased from eclipse/omr/master, the HEAD commit is generally volatile and in test.  

Until a new PR is submitted for this work source can be reviewed at https://github.com/eclipse/omr/compare/master...ktbriggs:evacuator. Please post review comments and discussion relating to evacuator source or performance here. Benchmarking results will be posted here at appropriate milestones.

**Comment by ktbriggs** [comment]
Baseline results for [WIP Evacuator: Baseline for 2664 development](https://github.com/ktbriggs/omr/commit/71115a24c2d02a713033c8ed481bbe0999b9242f) are below. Each benchmark was run 5 times with each gencon algorithm on an 8-core linux vm running rhel6 (rhel6x64cudavm1). Heap size was fixed at -Xms=-Xmx=1G for all runs, there were no global collections after the first nursery collection in any of these runs.

Performance metrics from gc logs.

|algorithm|benchmark|gc-count|gc-ms|gc-%cpu|gc-kb/ms|duration-ms|
|--------:|--------:|---:|-------:|----:|------:|-------:|
|scavenger|transactional|82176|17.516|99.764|1487|1439441|
|evacuator|transactional|83311|14.476|99.716|1797|1206052|
|scavenger|data compilation|6549|32.662|99.627|1494|213904|
|evacuator|data compilation|6629|27.200|99.535|1775|180312|
|scavenger|database processing|10889|1.560|96.861|1026|16992|
|evacuator|database processing|10768|1.427|96.082|1134|15366|

Perccent differences (100*(evacuator-scavenger)/scavenger) for nursery collection throughput (gc-kb/ms), duration (duration-ms) and benchmark throughput (scores) for these runs.

|benchmark|gc-kb/ms|duration-ms|score|
|--------:|---:|-------:|-----:|
|transactional|20.847%|-16.214%|1.422%|
|data compilation|18.809%|-15.714%|2.200%|
|database processing|10.526%|-9.569%|0.050%|

**Comment by ktbriggs** [comment]
Benchmarking results for [WIP Evacuator: Recursively scan root/remembered referent objects](https://github.com/ktbriggs/omr/commit/44552f4b6f5b3df41ef261c4ff045a7c40cf8f81) are below. Each benchmark was run 5 times with each gencon algorithm on an 8-core linux vm running rhel6 (rhel6x64cudavm1). Heap size was fixed at -Xms=-Xmx=1G for all runs, there were no global collections after the first nursery collection in any of these runs.

Performance metrics from gc logs.

|algorithm|benchmark|gc-count|gc-ms|gc-%cpu|gc-kb/ms|duration-ms|
|--------:|--------:|---:|-------:|----:|------:|-------:|
|scavenger|transactional|83015|17.837|99.769|1460|1480777|
|evacuator|transactional|83418|14.554|99.743|1791|1214085|
|scavenger|data compilation|6623|32.893|99.675|1481|217853|
|evacuator|data compilation|6773|26.994|99.531|1790|182834|
|scavenger|database processing|11112|1.518|97.544|1034|16873|
|evacuator|database processing|10912|1.499|96.217|1063|16364|

Percent differences (100*(evacuator-scavenger)/scavenger) for nursery collection throughput (gc-kb/ms), duration (duration-ms) and benchmark throughput (scores) for these runs.

|benchmark|gc-kb/ms|duration-ms|score|
|--------:|---:|-------:|-----:|
|transactional|22.671%|-18.010%|0.750%|
|data compilation|20.864%|-16.074%|3.787%|
|database processing|2.805%|-3.113%|1.955%|

This shows some improvement in benchmark throughput with evacuator as expected, without sacrificing gc throughput. The first run of the transactional benchmark produced a benchmark score that was ~25% lower than the mean, which may be due to environmental factors. Omitting this and the lowest scavenger score, there was a 2.557 improvement with evacuator vs scavenger. 

These results are corroborated by identical trials run on an 8-core linux vm running ubuntu 16.04 (fyre vm). From those trials, the percent differences in benchmark scores were 8.951%, 6.310% and 3.058% for the transactional, data compilation and database processing benchmarks, respectively. 

Further improvement in benchmark throughput is expected with the next commit, which will permit interleaving of survivor and tenure frames on the scan stack.

**Comment by ktbriggs** [comment]
Note that the `gc-%cpu` stats in the tables above are misleading. These are calculated for 8 gc threads from values in the verbose gc logs:

```
<gc-end ... type="scavenge" ... durationms=R usertimems=U systemtimems=S ...>
```

as (U+S)/(8*R)%. But U includes time CPU was stalled with no threads runnable as when evacuator stall, end, or synchronize. Since java was the only user process running, these voluntary wait states almost always stalled the CPU and parallelism stats reported in the gc logs are misleading.

More precise parallelization stats are maintained in the runtime scavenger stats, which evacuator also maintains, including counts and elapsed microseconds per gc thread for stall, end and synchronization events, but these are not reported in gc logs (yet). Evacuator builds with `EVACUATOR_DEBUG_ALWAYS` defined in EvacuatorBase.hpp enable reporting of these stats at the end of each generational collection for scavenger (default gencon algorithm) and evacuator (gencon + -Xgc:recursiveScanOrdering) operation. Additionally, histograms for object sizes, work packet sizes and cache sizes are reported along with counts for reference arc cache line containment (64/128/256 byte cache lines), etc. There is a minimal instrumentation overhead with these builds, which are otherwise optimized as for release builds.

I'm going to produce instrumented builds for the 2 commits reported above and redo the benchmarking runs on a more generic machine (not ...cuda...). Will report the results with more precise CPU utilization and cache line containment stats here.

Note: More detailed reporting for evacuator runs can be obtained by defining `EVACUATOR_DEBUG` in EvacuatorBase.hpp (this also enables 100s of assertions in evacuator code that incur substantial overhead). Debug evacuator builds recognize an environment variable `EVACUATOR_DEBUG_FLAGS` containing a bitmap of diagnostic categories to print to stderr when java is running with gencon/evacuator. To obtain output at end of generational gc as for `EVACUATOR_DEBUG_ALWAYS`, `export EVACUATOR_DEBUG_FLAGS=1`; see EvacuatorBase.hpp for other diagnostic flags.

**Comment by ktbriggs** [comment]
Work on interleaving survivor/tenure frames on the stack has revealed a couple of design flaws. I will finish and benchmark this commit and publish the results here, expect significant regression due to poor parallelism. 

The first flaw relates to the decision to flush frames from the top of the stack down in the presence of stalled evacuator threads. The frames at the top of stack are less dense with references than frames below, and the products of flushing them are even less dense and make light work for evacuators that pick them up. Instead, running evacuators with empty worklists should service stalled evacuator threads by moving work to their worklist from the bottom frames in pushed order until they accumulate some minimum volume of distributable work and notify. The contributing evacuator will just pop over these empty frames on the way down after the stall condition clears.

The second flaw relates to the choice of 256 bytes to restrict which objects can be copied inside the stack. Ideally, to maximize cache line containment, all objects that are not large pointer arrays (which must be split into segments for parallel scanning) should be scanned and copied within the stack. The minimum size of a split array segment (~4k) is a pivotal constant in this regard and should determine maximum object size for inside scanning and copying and frame copy limit.

I'll undertake these changes next after benchmarking interleaved frames with the warts in. 

With that said and two mea culpas in I'm going to lay on a little self-kudos for evacuator design. As evacuator takes shape I am finding it very workable and easy to make important changes such the two cases raised here. There are good reasons for this.

The main win comes from separating  evacuator (the generational collection task) from the collector (`MM_Scavenger` minus code reachable from `MM_Scavenger::workThreadGarbageCollect()`, the scavenger generational collection task). The collector's main concerns relate to whole-heap dynamics, including initiating and percolating collection cycles. Its nexus with the generational collection task is a single point that transfers control to 1 or more GC slave threads that effect the generational collection task. Separating these components simplifies and clarifies their implementations. 

Another win comes from the articulation of a family of specialized spaces, each minimally defining a contiguous range of addresses in survivor or tenure spaces. So evacuator has whitelists of whitespaces (reserved survivor or tenure space), a work list of workspaces (series of contiguous objects copied but not scanned), two outside copyspaces (space with a copy head, with work behind and white ahead), a large object copyspace (for receiving large objects to dispatch to the worklist), and a stack of scanspaces (copyspace with a scan head with completed work behind and work or whitepsace (pop!) ahead. Work and whitespace can be moved between different types of spaces with simple operators that effect the necessary pointer transforms.

Moreover, evacuator spaces allow optimal size determination for whitespace reservation and threshold for releasing work to workspace for inclusion on the worklist to be applied independently. This is in contrast to scavenger's `MM_CopyScanCacheStandard`, which makes work packet size dependent on the cache allocation size. This entanglement becomes a nuisance when, for example, scavenger attempts to increasing responsiveness to stalling threads by reducing work packet size.

**Comment by ktbriggs** [comment]
Further to my comment above regarding CPU usage stats reported in the GC logs, there is also a problem with the internal idle-time stats maintained in `MM_ScavengerStats`. These stats (stall, sync, completion counts and times) do not give a complete picture of total CPU time available and time utilized because they do not include time between point of thread dispatch and time thread actually kicks off and starts running. I think this value should be tracked in scavenger stats, and that the endpoints for the cycle should be the time the first thread enters its main method and the time the last thread leaves its main method. Also I think these stats should be normalized per the %cpu allocated to the java process as currently reported in the `<gc-end ... type="scavenge" ... durationms=R usertimems=U systemtimems=S ...>` clauses in GC logs and reported in the GC log. I will create an issue for this.

**Comment by ktbriggs** [comment]
Benchmarking results for [WIP Evacuator: Interleave survivor/tenure scanspaces ](https://github.com/ktbriggs/omr/commit/1b9ce56813a462f13a39ff96b13b41f180abe06c) are below. Each benchmark was run 5 times with each gencon algorithm on an 8-core linux vm running rhel6 (lnxem64tvm7-resv). Heap size was fixed at -Xms=-Xmx=1G for all runs to minimize global collections.

Note that these benchmarks were run on a different machine than that used to benchmark the last two commits. Also, this build is instrumented so that scavenger and evacuator print a summary report after each nursery collection. This report contains more detailed information and I will present these data after presenting results that can be compared with the results reported for the previous two commits, bearing in mind the machine difference.

Performance metrics from gc logs.

|algorithm|benchmark|gc-count|gc-ms|gc-%cpu|gc-kb/ms|duration-ms|
|--------:|--------:|---:|-------:|----:|------:|-------:|
|scavenger|transactional|44643|14.825|99.396|1779|661866|
|evacuator|transactional|43763|24.952|95.753|1061|1092003|
|scavenger|data compilation|6642|29.860|99.496|1962|198335|
|evacuator|data compilation|6152|40.609|97.553|1447|249831|
|scavenger|database processing|10068|2.679|94.329|1010|26981|
|evacuator|database processing|10119|5.162|92.324|522|52239|

Percent differences (100*(evacuator-scavenger)/scavenger) for nursery collection throughput (gc-kb/ms), duration (duration-ms) and benchmark throughput (scores) for these runs.

|benchmark|gc-kb/ms|duration-ms|score|
|--------:|---:|-------:|-----:|
|transactional|-40.359%|64.988%|-1.943%|
|data compilation|-26.248%|25.964%|-7.128%|
|database processing|-48.316%|93.614%|-1.000%|

Evacuator gc throughput and duration have regressed markedly relative to the previous commit. Benchmark scores have also regressed in the evacuator runs. These results were anticipated because this mode of stack operation sends only large objects (>256 bytes) to outside copyspaces, which accumulate unscanned work to release to the evacuator worklist as distributable workspaces. Previous to this commit the stack was homogeneous and limited copy inside to small objects within the same region; objects forwarded across the nursery/tenure boundary were always copied outside. With interleaved survivor/tenure frames all small objects are scanned and copied inside the stack on the stack and the volume of outside copy is reduced. 

Instrumented metric data for scavenger/evacuator for these benchmark runs show that evacuator threads are idle for substantial periods of time during collections and that running evacuator threads are flushing large volumes of material to outside copyspaces. Evacuator threads respond to the presence of stalled evacuator threads by lowering the threshold for releasing accumulated work from outside copyspaces to a minimum and flushing stack frames from the top down to the outside copyspaces until the stall condition clears. The stall condition is checked after every slot is scanned, so flushing evacuators produce a minimal volume of outside work and any distributable workspaces released in the process are froth from the top of evacuator stacks and so are less dense with recursive work. The stalled evacuators that pick these workspaces up make short work of them and stall again, with the overall result that there is a lot stalling with waylotsa traffic on the work distribution bus and correspondingly more system overhead (this is apparent in the `gc-%cpu` stats above). 

That's not good, but it can be fixed. I'm going to walk my dog, and when I come back I'll illustrate my point above with some charts from the scavenger/evacuator instrumentation data.

**Comment by ktbriggs** [comment]
Subsequent to benchmarking trials above javac failed on a couple of assertions relating to changes to stack whitespace management, which was reworked for _Interleave survivor/tenure scanspaces _. Also, 32-bit compilation failed on all supported platforms on an attempt to coerce a 64-bit integer into a static const declarration. So there will be at least one more commit squashed into this one. I'm including results and discussion for this commit only to show the effect of interleaving on evacuator parallelism and stack throughput.  The wrinkles will be ironed out in subsequent commits (see design note above).

**Comment by ktbriggs** [comment]
Parallelism metrics for [WIP Evacuator: Interleave survivor/tenure scanspaces ](https://github.com/ktbriggs/omr/commit/1b9ce56813a462f13a39ff96b13b41f180abe06c). 

These stats are derived from instrumentation data recorded in `MM_ScavengerStats`  for scavenger and evacuator runs. These data include total counts and summed wait times for stall, complete and sync events per gc thread per nursery collection and real time duration for each collection. A missing component is gc thread start time (between thread dispatch and entry into main method).

|algorithm|benchmark|%stall|%complete|%sync|%cpu|%inside|
|---------:|---------:|------:|---------:|-----:|----:|-------:|
|scavenger|transactional|7.355|4.465|1.188|86.991|46.833|
|evacuator|transactional|93.050|5.068|0.650|1.231|38.376|
|scavenger|data compilation|5.038|18.108|0.758|76.094|92.847|
|evacuator|data compilation|68.340|4.554|0.444|26.661|25.669|
|scavenger|database processing|21.023|25.216|23.671|30.089|64.190|
|evacuator|database processing|67.546|**58.346**?|12.922|**19.53**? |38.376|

There is obviously a problem with the `%complete` value calculated for evacuator/database processing, which I can't explain yet so I've omitted it from the `%cpu` calculation for this row. I think the `%cpu` values (`100 - (%stall + %complete + %sync)`) are about right, and the general picture is clear. 

The `%inside` values are a bit apples and oranges. For evacuator this is the percentage of evacuated volume that was copied inside the stack. For scavenger this is the percentage of evacuated volume that was pulled from the worklist (excluding deferred caches) for scanning. This is approximate to the percentage scanned and copied within aliased caches. 

Looking at the transactional benchmark, which presents >97% small objects to the collector, the `%inside` value (~38%) is less than half what would be expected (~97%) with no stalling. There were ~10 times as many stall events and the average duration was ~2 times longer, compared to scavenger, and cpu utilization was on the floor. More than 60% of all objects small or large were flushed to outside copyspaces to service stalled threads and, importantly I think, most of the flushed objects encapsulated only small volumes of evacuation volume because they were skimmed from the top of the stack.

So this will have to be improved. I'm going to try out the ideas outlined in the design note above and brb.

**Comment by ktbriggs** [comment]
Benchmarking results for [WIP Evacuator: Interleave survivor/tenure scanspaces (instrumented)](https://github.com/ktbriggs/omr/commit/2b0a9786beddbc2226060442d426290499071be5) are below. Each benchmark was run 5 times with each gencon algorithm on an 8-core linux vm running rhel6 (rhel6x64cudavm1). Heap size was fixed at -Xms=-Xmx=1G for all runs.

Changes included in this commit were:

- Allow small object (<=256b) push() from survivor to tenure and vice versa on the evacuator scan stack. 
- Flush frames from bottom of stack up to and excluding current frame in presence of stalled evacuators.
- Simplify loading of root objects into scan stack.
- Enable spin for monitor-entry, disable for monitor-try-enter.
- Stop coalescing workspaces when adding to worklist.

Performance metrics from gc logs.

|algorithm|benchmark|gc-count|gc-ms|gc-%cpu|gc-kb/ms|duration-ms|
|--------:|--------:|---:|-------:|----:|------:|-------:|
|scavenger|transactional|40327|19.391|99.793|1340|782000|
|evacuator|transactional|41627|16.262|99.792|1598|676976|
|scavenger|data compilation|5345|45.274|99.873|1364|241990|
|evacuator|data compilation|5573|37.022|99.796|1669|206324|
|scavenger|database processing|8618|1.723|97.784|1068|14854|
|evacuator|database processing|8999|1.622|96.677|1118|14601|

Percent differences (100*(evacuator-scavenger)/scavenger) for nursery collection throughput (`gc-kb/ms`), duration (`duration-ms`) and benchmark throughput (scores) for these runs.

|benchmark|gc-kb/ms|duration-ms|score|
|--------:|---:|-------:|-----:|
|transactional|19.253%|-13.430%|4.513%|
|data compilation|22.361%|-14.739%|3.787%|
|database processing|4.682%|-1.703%|4.711%|

This build was instrumented so that evacuator and scavenger output additional internal metrics derived from `MM_ScavengerStats` per nursery cycle. These provide a more detailed and accurate picture of the percentage of available cpu resources used by evacuator and scavenger threads during nursery collections as well as statistics relating to the prevalence of small objects (<=256b), proportion of scanned material that was generated and consumed inside aliased caches (scavenger) or inside scan stack frames (evacuator), as opposed to pulled from outside (ie, from work queues), and the percentages of object evacuations that resulted in cache line containment for the reference arc (64-, 128- and 256-byte cache lines as for x86-64, powerpc, and z-series platforms, respectively). The important difference between `gc-%cpu` as derived from verbose gc logs in the table above and `gc-%cpu` reported for the internal metrics is that the former includes time spent with active thread spinning (on a spin lock) and time wasted with cpu allocated but idle -- if jvm is the only user process running on the platform (as for these benchmarking runs) then this statistic will always be close to 100%. The `gc%cpu` stats derived from internal metrics excludes these wait times.

Performance metrics from internal metrics (cache line containment percentages are cumulative).

|algorithm|benchmark|gc-%cpu|%small|%inside|%<64b|%<128b|%<256b|
|--------:|--------:|---:|-------:|----:|------:|-------:|-------:|
|scavenger|transactional|97.469|97.357|73.442|6.218|9.791|16.579|
|evacuator|transactional|98.803|97.389|90.439|12.626|18.031|26.165|
|scavenger|data compilation|96.150|77.341|93.199|2.878|5.971|11.853|
|evacuator|data compilation|96.396|77.387|70.348|8.989|17.753|29.022|
|scavenger|database processing|76.808|51.638|75.573|12.090|14.064|15.481|
|evacuator|database processing|73.249|52.273|29.720|16.321|19.237|21.613|

**Comment by ktbriggs** [comment]
**[Incomplete, waiting on (64,4108) configuration runs to complete]**

Benchmarking results for [WIP Evacuator: Options for evacuator tuning (instrumented)]{https://github.com/ktbriggs/omr/commit/0b35e5c551636fb3484c482786138006dcc002f3} are below. Each benchmark was run 5 times with each gencon algorithm on an 8-core linux vm running rhel6 (rhel6x64cudavm1). Heap size was fixed at -Xms=-Xmx=1G for all runs.

Changes included in this commit are:

- add jvm option `-XXgc:recursiveMaximumInsideCopyDistance` -- allow copy inside current frame as long as the span between scan and copy heads not greater than this value, otherwise push and copy to next frame. Default 0 bytes, which forces depth-first scanning and copying on the stack. Allowable values are in the in closed range [0, 768]. 
- add jvm option `-XXgc:recursiveMaximumInsideCopySize` -- allow objects up to this size in bytes to be copied inside stack whitespace. Default 4108 or 8216 bytes (compressed or uncompressed references), allow values in closed range [64,4108|8216]. The upper limit corresponds to the size in bytes of the smallest pointer array that can be split for scanning with compressed|noncompressed fomrobject_t format.

The default copy span `recursiveMaximumInsideCopyDistance=0` forces depth-first scanning and copying inside the stack. The default copy size `recursiveMaximumInsideCopySize=4108` admits almost all objects except splitable arrays for copying inside the stack for optimal performance. I ran the 5X benchmark suite on rhel6x64cudavm1 with default `recursiveMaximumInsideCopySize=4108` for three `recursiveMaximumInsideCopyDistance` configurations: 

-`recursiveMaximumInsideCopyDistance=0`: scan and copy depth-first always
-`recursiveMaximumInsideCopyDistance=64`: scan and copy inside frame until scan-copy span >64 bytes then depth-first until end of frame
-`recursiveMaximumInsideCopyDistance=768`: scan and copy inside frame until scan-copy span >768 bytes then depth-first until end of frame

I am reporting only the most salient outcomes for these runs, carrying forward outcomes from [WIP Evacuator: Interleave survivor/tenure scanspaces (instrumented)](https://github.com/ktbriggs/omr/commit/4e51119f78dcba89efa7cbfc6f27230cdfd6ef85) for comparison (these correspond roughly to a (768,256) configuration for (span,size) and are labeled (-,256) in the table below.

|algorithm|benchmark|span|size|gc-kb/ms|%inside|%<64b|%<128b|%<256b|
|--------:|--------:|-------:|----:|-------:|------:|----:|-----:|-----:|
|scavenger|transactional|-|256|1340|73.442|6.218|9.791|16.579|
|evacuator|transactional|-|256|1598|90.439|12.626|18.031|26.165|
|evacuator|transactional|0|4108|1379|92.399|28.486|36.083|40.562|
|evacuator|transactional|64|4108|na|na|na|na|na|
|evacuator|transactional|768|4108|1531|92.247|19.174|25.242|31.738|
|scavenger|data compilation|-|256|1364|93.199|2.878|5.971|11.853|
|evacuator|data compilation|-|256|1669|70.348|8.989|17.753|29.022|
|evacuator|data compilation|0|4108|1406|84.079|28.758|33.035|36.405|
|evacuator|data compilation|64|4108|na|na|na|na|na|
|evacuator|data compilation|768|4108|1638|85.706|7.949|16.218|28.001|
|scavenger|database processing|-|256|1068|75.573|12.090|14.064|15.481|
|evacuator|database processing|-|256|1118|29.720|16.321|19.237|21.613|
|evacuator|database processing|0|4108|1071|31.983|19.502|22.099|23.837|
|evacuator|database processing|64|4108|na|na|na|na|na|
|evacuator|database processing|768|4108|1123|31.751|16.386|19.324|21.887|

Percent differences (100*(evacuator-scavenger)/scavenger) for benchmark throughput (scores) for these runs.

|benchmark|span|size|score|
|--------:|-------:|----:|----:|
|transactional|-|256|4.513|
|transactional|0|4108|2.369|
|transactional|64|4108|na|
|transactional|768|4108|2.368|
|data compilation|-|256|4.711|
|data compilation|0|4108|0.456|
|data compilation|64|4108|na|
|data compilation|768|4108|5.267|
|database processing|-|256|0.647|
|database processing|0|4108|0.529|
|database processing|64|4108|na|
|database processing|768|4108|0.062|

**Comment by ktbriggs** [comment]
#### Aggregate scavenger benchmarking results (rhel6x64cudavm1)

|benchmark|gc-count|gc-kb/ms|%inside|%<64b|%<128b|%<256b|
|-----------:|---------:|----------:|-------:|------:|--------:|---------:|
|transactional|4809|1349|73.138|6.214|9.784|16.567|
|data compilation|1052|1397|93.265|2.874|5.969|11.861|
|database processing|643|1079|75.972|12.113|14.089|15.503|

All values in the table are averages, N=25.

**Comment by ktbriggs** [comment]
I am attaching some images that may help to illustrate how evacuator works. These describe evacuator circa _WIP Evacuator: Interleave survivor/tenure scanspaces (instrumented)_. 


![heap](https://user-images.githubusercontent.com/18146124/45658116-4e4b2b80-babb-11e8-964d-ad49b3fcc34e.PNG)

![stack](https://user-images.githubusercontent.com/18146124/45658135-69b63680-babb-11e8-91bb-404258b98dc8.PNG)

![evacuator](https://user-images.githubusercontent.com/18146124/45658140-7044ae00-babb-11e8-8944-8330482d6469.PNG)

![transducer](https://user-images.githubusercontent.com/18146124/45658146-789ce900-babb-11e8-83d3-a51b63a593e0.PNG)

**Comment by ktbriggs** [comment]
Depth-first scan stack operation if a WIP. I've taken a first run at optimizing the scanning loop for depth first and initial results are disappointing but provide some insight into evacuator bottlenecks. I instrumented the stack to count frame activation (frame is activated when an object is pushed into the frame or a workspace is pulled into the bottom frame). Below is a histogram showing average proportions of activation per frame per nursery collection for the three benchmarks I've been tracking.

![image](https://user-images.githubusercontent.com/18146124/45832904-2df4ba00-bcd1-11e8-81f4-5901dfa42fc6.png)

Those are relative to the overall volume of material that is copied inside the stack versus copied to outside copyspaces, as shown below.

![image](https://user-images.githubusercontent.com/18146124/45833048-a5c2e480-bcd1-11e8-88dd-725617e574b1.png)

The data compilation benchmark presents a lot of linked structures, which force material to outside copy when they overflow the stack. The database processing benchmark I think is presenting a large number of array structures, which force outside copying whenever the remainder stack whitespace is to small to receive an object >32 bytes in size. in those cases the optimized path for copying inside is less frequently used and evacuation slows down.

**Comment by ktbriggs** [comment]
##### Evacuator only, depth-first scanning, N=10

|scan-gb|%tenured|scan-kb/ms|%cpu|%inside|%<64b|%<128b|%<256b|
|------:|-------:|---------:|---:|------:|----:|-----:|-----:|
|_transactional_|
|199.126|.182|1317.957|99.982|89.482|45.779|57.722|64.704|
|_data_|_compilation_|
|63.841|14.936|1376.604|99.965|85.809|45.964|52.793|58.178|
|_database_|_processing_|
|3.045|14.241|941.811|99.609|35.107|21.203|24.132|26.137|

Percentage difference in total benchmark throughput (scores) per trial (100*(evacuator-scavenger)/scavenger)

|transactional|data compilation|database processing|
|--:|--:|--:|
|0.601|0.703|0.489|

**Comment by ktbriggs** [comment]
Here are typical verbose gc plots showing nursery size and volumes of live material evacuated to survivor and tenure spaces for the three benchmarks included in these trials. The tilt ratio is the volume of the evacuation space as a percentage of the nursery size. The amounts flipped and tenured are the volumes evacuated to survivor and tenure space, respectively.

These are from evacuator trails with the most recent (depth-first) commit.

#### Transactional (tilt ratio ~89%)

![image](https://user-images.githubusercontent.com/18146124/45946320-950abb00-bfbe-11e8-87ec-6e6485823d1a.png)

#### Data compilation(tilt ratio ~77%)

![image](https://user-images.githubusercontent.com/18146124/45946397-cbe0d100-bfbe-11e8-93ea-f515295dc4ad.png)

#### Database processing (tilt ratio ~90%)

![image](https://user-images.githubusercontent.com/18146124/45946466-f7fc5200-bfbe-11e8-82a0-8febdc7f1a63.png)

The database processing benchmark at steady state will never show any appreciation with increasing cache line containment as almost all live objects die early and are never evacuated or die early in the next generation if they are flipped. In any case cache line containment is short-lived as reference arcs die when either endpoints dies. The high volume of evacuation seen in the first two minutes relates to loading the database into RAM -- this is time not factored into the final benchmark score, which reflects performance during the last four minutes at steady state. Cache line containment obtained when live objects are tenured is persistent in the heap image of the database, but benchmark throughput is indifferent to this as a determining factor. 

The data compilation benchmark shows pressure on the nursery, with transient live objects frequently overflowing into tenure space and dying soon after. Each trial run of scavenger or evacuator with this benchmark incurred ~6 global collections of tenure space. I would expect that this benchmark would be more sensitive cache line containment as a determinant for benchmark throughput if it were run with a larger nursery. This benchmark showed marked increase in throughput relative to scavenger and other evacuator configurartions in the trials reported above for _Evacuator only, recursiveMaximumInsideCopyDistance=768, N=5_ -- this configuration allowed recursive structures to be spun out within a stack frame indefinitely as long as the reference arc length for object copy remained <768 bytes. But it is unclear whether this was due to evacuator throughput or cache line containment as both of these factors improved in this configuration.

**Comment by ktbriggs** [comment]
I have reset the **evacuator** branch in my github and local OMR repos to the _WIP Evacuator: Interleave survivor/tenure scanspaces (instrumented)_ commit and similarly reset the evacuator branch in my github and local OpenJ9 repos to the contemporaneous _WIP Evacuator: Recursively scan root/remembered referent objects_ commit. 

These commit correspond to the build that has produced the best results on the test benchmarks and I am reporting these results publicly in a poster expo at CASCON 2018 along with cloning coordinates for these stable evacuator branches and for OpenJDK in case anyone wants to build it themselves. 

Follow-on commits reported here, and ongoing evacuator work, have been moved to the **evacuator-wip** branches in my github and local OMR and OpenJ9 repos.

**Comment by ktbriggs** [comment]
> PR

equ

**Comment by ktbriggs** [comment]
I am capping work on evacuator at this point. Most of the viable experiments outlines in the first comment above have been concluded, with the exception of reversing order of mutator thread stack walk (currently top-down, my conjecture is that bottom-up traversal would improve parallelism...). Most of the work done since the last report has been focused on optimizing and streamlining evacuator scan stack operation to improve depth-first performance, conserve whitespace, and reduce compiled code size with more judicious inlining of methods. I will be updating the evacuator design (#2361) to reflect changes and new features introduced since it was written several months ago and will include discussion to interpret results reported here. 

I ran a comprehensive suite of benchmarks, including the transactional, data compilation, and database processing benchmarks that were run against earlier iterations . I have added another, more processing-intensive benchmark that I refer to as _enterprise_ in the tables below. 

![image](https://user-images.githubusercontent.com/18146124/52053349-223ebd00-2527-11e9-810e-21e328b24842.png)

The enterprise benchmark was run twice for each run configuration and the results reported above are averages over the two runs. This benchmark takes about 2 hours to run and produces two scores: score-A is sensitive to application throughput, score-B is sensitive to GC throughput. The enterprise runs were executed on a single NUMA node (node 0 of lnxem64tvm10-resv.ottawa.ibm.com, 30 cores, 64gb RAM) using a 56gb heap with default (15.5gb nursery).

The other benchmarks were run 5 times in each run configuration and these results are averaged over the 5 runs. These were executed on a single NUMA node (node 0 of lnxem64tvm6-resv.ottawa.ibm.com, 8 cores, 9gb RAM). Allocated heap/nursery sizes were as for previously reported  runs (transactional 2048/512mb, data compilation 1280/640mb, data processing 1280/320mb).

**Comment by ktbriggs** [comment]
With some tweaking to improve whitelist utilization, and to increase the overflow threshold from 3 to 8 for trimming oversized remainder whitespace from copyspaces, obtained the results shown below. Whitespace remainders in scan stack frames are required to be <32 bytes before they can be trimmed and the remainder discarded. In previous builds objects &ge;32 bytes that would normally be copied inside the stack would be copied to an outside copyspace if they did not fit in the remaining whitespace. For outside copyspaces whitespace remainders were never trimmed until they were reduced to <768 bytes or the overflow threshold (8*minimum workspace size, remainders &ge;768 bytes are trimmed if more than this amount of material overflowed the copyspace) was breached. 

In this build, if an object less than the maximal inside copy size does not fit in the stack whitespace remainder and the whitelist holds a larger whitespace fragment that can accommodate the object, the stack whitespace remainder is trimmed and swapped into the whitelist and the top of the whitelist is swapped into the stack to receive the object. This has the effect of reducing the whitelist volume as desired and allowing inside copy to continue, Similarly, if an outside copyspace is in an overflow state with &ge;768 whitespace bytes remaining the remainder can be trimmed and swapped into the whiitelist if the top (largest) fragment on the whitelist can receive the object. 

The _enterprise_ benchmark was run 3x2 times interleaving evacuator and scavenger on lnxem64tvm10-resv.ottawa.ibm.com configured as for the January 31 2019 trials reported above. The other three benchmarks were run 5x2 times interleaving evacuator and scavenger on lnxem64tvm7-resv.ottawa.ibm.com configured as for  January 31 2019 trials. This machine has the same hardware architecture and NUMA configuration as lnxem64tvm6-resv.ottawa.ibm.com (modulo different RAM sizes not material to these trials), which was used for these benchmarks in the January 31 trials.

![image](https://user-images.githubusercontent.com/18146124/53181957-8b659d80-35c6-11e9-98a0-1024e412b5ab.png)

I've included the `inside%` stats for evacuator. This metric measures the percentage of objects that are copied inside stack frames vs copied to outside copyspaces, Note that improvement in benchmark primary scores (`score-A%`) tends to be directly if not linearly related to evacuator `inside%` (correlation coefficient 0.83) and inversely related to evacuator `cache-64%` (correlation coefficient -0.2835). This suggests that the relative volume of inside copying is a stronger determinant of benchmark performance than is cache-line containment. A likely explanation for this is that evacuator threads are able to discover new scan work as they recurse into objects copied ahead within stack frames and are able to perform a greater proportion of work without drawing from their own or other evacuator worklists. 

I think that a similar process explains the advantage of scavenger's hierarchical copying vs breadth-first copying. Scavenger's deferred scancache acts in a similar manner to evacuator's bottom stack frame. When a scavenger thread runs out of scan work it will pull in the deferred scancache and start scanning when it previously left off, but will likely alias to the scancache that receives the first object that it copies and commence scanning hierarchically after deferring the pulled scancache again. In a nutshell, scavenger threads perform a pseudo-random walk that covers the intersection of the reference graph and evacuation space, discovering new work in the process and pulling from worklists only when the deferred scancache and both (survivor and tenure) local scancaches have been completely scanned. 

Scavenger's deferred scancache may be displaced to the worklist if another scancache must be deferred, which means that any scancache can be cycled through the worklist circuit multiple times in the process of being scanned from base to end. So the `inside%` metric is difficult to measure for scavenger, because scancaches can appear >1 times on worklists, with less scan work available on each iteration. Where it is reported for previous trials it is overestimated to a greater or lesser extent.

**Comment by ktbriggs** [comment]
Here are final results from evacuator vs scavenger benchmarking with most recent builds. The material changes since the last report (Feb 21, above) relate to whitespace management. In these builds I have reverted the changes described in Feb 21 comment, reduced the tolerance for trimming remainder whitespace from outside copyspaces from 768 to 256 bytes, and permitted objects that overflow outside copyspaces to be redirected into stack whitespace if they fit in current stack whitespace. I have also introduced 3 flush conditions, which force all objects to be copied to outside whitespaces whenever one or more flush condition is present. The 3 flush conditions are reevaluated before each object evacuation. They are:

1- presence of one or more stalled evacuators
2- outside copyspace overflow (object size > copyspace remainder > 256 bytes)
3- stack overflow 

The last flush condition forces depth-first stack operation while the stack grows and forces flushing to outside copyspaces as the stack winds down. This is intended to help parallelize evacuation of recursive structures with long recursing chains of objects such as might be encountered in, for example, XML/DOM processing contexts. 

I report results from 3 full sets of benchmarking runs below. Each set was run on the same machines used in previously reported benchmarking runs eg, Feb 21. Each run used a different build, but the changes between these builds were not material to scavenger or evacuator performance -- there were no substantial changes to scavenger and changes to evacuator related only to some cleanup work to rename a class and remove a couple of unused artifacts. All of these builds were internal Linux/x86_64 builds for J9 Java VM (80).

**_Build jvmxa6480-408175:_**

![image](https://user-images.githubusercontent.com/18146124/55007523-efa8c200-4fb5-11e9-9bc8-5ae515a59ec3.png)

**_Build jvmxa6480-410164:_**

![image](https://user-images.githubusercontent.com/18146124/55008197-056ab700-4fb7-11e9-9d72-746b0a894e33.png)

**_Build jvmxa6480-412811:_**

![image](https://user-images.githubusercontent.com/18146124/55008549-845fef80-4fb7-11e9-8c29-86da55dcd08f.png)

Build 412811 was produced from the `evacuator-wip` branches of my personal omr and openj9 repos at https://github.com/ktbriggs/omr/tree/evacuator-wip and https://github.com/ktbriggs/openj9/tree/evacuator-wip. These branches differ from the stable `evacuator` branches in the respective repos only insofar as the `...-wip` branches are instrumented to produce summary output per nursery collection, instrumentation is disabled in the stable `evacuator` branches. The only source code difference between the `evacuator-wip` and `evacuator` branches is that in `omr/evacuator-wip`

```
#undef EVACUATOR_DEBUG_ALWAYS
```

is replaced with 

```
#define EVACUATOR_DEBUG_ALWAYS
```

in `omr/gc/base/Evacuator.hpp`. The format and content of instrumentation output is described in #2361.

As things stand evacuator is stable and working as designed in builds for Linux/x86_64 and in 64-bit builds for other platforms. There is a bug in 32-bit builds (all platforms) that I believe relates to evacuator interactions with recent scavenger changes (this bug is not present in eclipse/{openj9-omr,openj9} builds). I will not have an opportunity to debug this or perform further work on OpenJ9/OMR in the foreseeable future.  

I am closing this and #2361 as of now.
